rest/route/pod_agent.go

```diff
@@ -115,13 +121,91 @@ func (h *podAgentNextTask) Parse(ctx context.Context, r *http.Request) error {
 }
 
 func (h *podAgentNextTask) Run(ctx context.Context) gimlet.Responder {
-	j := units.NewPodTerminationJob(h.podID, "reached end of pod lifecycle", utility.RoundPartOfMinute(0))
-	if err := amboy.EnqueueUniqueJob(ctx, h.env.RemoteQueue(), j); err != nil {
+	p, err := h.findPod()
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+
+	h.setAgentFirstContactTime(p)
+
+	pd, err := h.findDispatcher()
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(err)
+	}
+
+	nextTask, err := pd.AssignNextTask(ctx, h.env, p)
+	if err != nil {
 		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
 			StatusCode: http.StatusInternalServerError,
-			Message:    errors.Wrap(err, "enqueueing job to terminate pod").Error(),
+			Message:    errors.Wrap(err, "dispatching next task").Error(),
 		})
 	}
+	if nextTask == nil {
+		j := units.NewPodTerminationJob(h.podID, "reached end of pod lifecycle", utility.RoundPartOfMinute(0))
+		if err := amboy.EnqueueUniqueJob(ctx, h.env.RemoteQueue(), j); err != nil {
+			return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+				StatusCode: http.StatusInternalServerError,
+				Message:    errors.Wrap(err, "enqueueing job to terminate pod").Error(),
+			})
+		}
+	}
```

Comment:
The empty struct is equivalent to returning 200 OK and telling the agent that there's no next task to run. Since the agent doesn't actually run tasks, this is supposedly going to be filled in with real logic in PM-2616.